File: ts/src/identityKey.ts
import { ml_dsa65 } from "@noble/post-quantum/ml-dsa";
import { arrayBufferToBase64, base64ToArrayBuffer, keyHash } from "./utils.ts";
import { isValidUUIDv7 } from "./core.ts";
import { isValidMasterKeyPrivate, isValidMasterKeyPublic, signMasterKey } from "./masterKey.ts";
import type { identityKey, Sign } from "./type.ts";

export function signIdentityKey(
  key: string,
  data: string,
  pubKeyHash: string,
): string | null {
  if (!isValidIdentityKeyPrivate(key)) {
    return null;
  }
  const { key: keyBinary } = JSON.parse(key);
  const dataArray = new TextEncoder().encode(data);
  const signature = ml_dsa65.sign(
    new Uint8Array(base64ToArrayBuffer(keyBinary)),
    dataArray,
  );
  const signString = arrayBufferToBase64(signature as unknown as ArrayBuffer);
  const signResult: Sign = {
    signature: signString,
    keyHash: pubKeyHash,
    keyType: "identityKey",
    algorithm: "ML-DSA-65",
  };
  return JSON.stringify(signResult);
}

export function verifyIdentityKey(
  key: string,
  sign: string,
  data: string,
): boolean {
  if (!isValidIdentityKeyPublic(key)) {
    return false;
  }
  const { key: keyBinary } = JSON.parse(key);
  const signData: Sign = JSON.parse(sign);
  if (signData.keyType !== "identityKey") {
    return false;
  }
  const dataArray = new TextEncoder().encode(data);
  const verify = ml_dsa65.verify(
    new Uint8Array(base64ToArrayBuffer(keyBinary)),
    dataArray,
    new Uint8Array(base64ToArrayBuffer(signData.signature)),
  );
  return verify;
}

export async function generateIdentityKey(uuid: string, masterKey: {
  publicKey: string;
  privateKey: string;
}): Promise<
  {
    publickKey: string;
    privateKey: string;
    sign: string;
  } | null
> {
  if (!isValidUUIDv7(uuid)) {
    return null;
  }
  if (!isValidMasterKeyPrivate(masterKey.privateKey)) {
    return null;
  }
  if (!isValidMasterKeyPublic(masterKey.publicKey)) {
    return null;
  }
  const seed = crypto.getRandomValues(new Uint8Array(32));
  const key = ml_dsa65.keygen(seed);
  const publicKeyBinary = arrayBufferToBase64(
    key.publicKey as unknown as ArrayBuffer,
  );
  const privateKeyBinary = arrayBufferToBase64(
    key.secretKey as unknown as ArrayBuffer,
  );
  const timestamp = new Date().getTime();
  const publickKeyObj: identityKey = {
    keyType: "identityKeyPublic",
    key: publicKeyBinary,
    algorithm: "ML-DSA-65",
    timestamp: timestamp,
    sessionUuid: uuid,
  };
  const privateKeyObj: identityKey = {
    keyType: "identityKeyPrivate",
    key: privateKeyBinary,
    algorithm: "ML-DSA-65",
    timestamp: timestamp,
    sessionUuid: uuid,
  };
  const publickKey = JSON.stringify(publickKeyObj);
  const privateKey = JSON.stringify(privateKeyObj);
  const sign = await signMasterKey(
    masterKey.privateKey,
    publickKey,
    await keyHash(masterKey.publicKey),
  );
  if (!sign) {
    return null;
  }
  return {
    publickKey: publickKey,
    privateKey: privateKey,
    sign: sign,
  };
}

export function isValidIdentityKeyPrivate(key: string): boolean {
  if (key.length !== 5496) {
    console.log(key.length);
    return false;
  }
  const { key: keyBinary, keyType } = JSON.parse(key);
  if (keyType !== "identityKeyPrivate") {
    return false;
  }
  const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
  if (keyBinaryArray.length !== 4032) {
    console.log(keyBinaryArray.length);
    return false;
  }
  return true;
}

export function isValidIdentityKeyPublic(key: string): boolean {
  if (key.length !== 2723) {
    console.log(key.length);
    return false;
  }
  const { key: keyBinary, keyType } = JSON.parse(key);
  if (keyType !== "identityKeyPublic") {
    return false;
  }
  const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
  if (keyBinaryArray.length !== 1952) {
    console.log(keyBinaryArray.length);
    return false;
  }
  return true;
}

export function isValidSignIdentityKey(sign: string): boolean {
  const { keyHash, signature, keyType } = JSON.parse(sign);
  if (sign.length !== 4509) {
    console.log(sign.length);
    return false;
  }
  if (keyType !== "identityKey") {
    return false;
  }
  if (keyHash.length !== 44) {
    console.log(keyHash.length);
    return false;
  }
  if (signature.length !== 4412) {
    console.log(signature.length);
    return false;
  }
  return true;
}

File: rust/src/identity_key.rs
use crate::r#type::IdentityKey;
use crate::signature::{create_signature_object_mlds65, verify_with_mlds65};
use crate::keyutils::generate_dsa65_key_pair;
use crate::core::is_valid_uuid_v7;
use crate::master_key::{is_valid_master_key_private, is_valid_master_key_public, sign_master_key};
use crate::utils::key_hash;
use chrono::Utc;
use serde_json;

/// 秘密鍵で IdentityKey に署名
pub fn sign_identity_key(
    key_json: &str,
    data: &str,
    key_hash: &str,
) -> Option<String> {
    let ik: IdentityKey = serde_json::from_str(key_json).ok()?;
    if ik.key_type != "identityKeyPrivate" { return None; }
    create_signature_object_mlds65(&ik.key, data.as_bytes(), key_hash, "identityKey").ok()
}

/// 公開鍵で IdentityKey の署名検証
pub fn verify_identity_key(
    key_json: &str,
    sign_json: &str,
    data: &str,
) -> bool {
    let ik: IdentityKey = match serde_json::from_str(key_json) {
        Ok(v) => v,
        Err(_) => return false,
    };
    if ik.key_type != "identityKeyPublic" { return false; }
    let sign: crate::r#type::Sign = match serde_json::from_str(sign_json) {
        Ok(v) => v,
        Err(_) => return false,
    };
    if sign.key_type != "identityKey" { return false; }
    verify_with_mlds65(&ik.key, data.as_bytes(), &sign.signature)
}

/// IdentityKey を生成し、マスター鍵で署名
pub fn generate_identity_key(
    uuid: &str,
    master_public_json: &str,
    master_private_json: &str,
) -> Option<(String, String, String)> {
    if !is_valid_uuid_v7(uuid) { return None; }
    if !is_valid_master_key_private(master_private_json) { return None; }
    if !is_valid_master_key_public(master_public_json)  { return None; }
    let (pub_b64, priv_b64) = generate_dsa65_key_pair().ok()?;
    #[cfg(target_arch = "wasm32")]
    let timestamp = 0u64;
    #[cfg(not(target_arch = "wasm32"))]
    let timestamp = Utc::now().timestamp_millis() as u64;
    let pub_obj = IdentityKey {
        key_type: "identityKeyPublic".into(),
        key: pub_b64.clone(),
        algorithm: "ML-DSA-65".into(),
        timestamp,
        session_uuid: uuid.into(),
    };
    let priv_obj = IdentityKey {
        key_type: "identityKeyPrivate".into(),
        key: priv_b64.clone(),
        algorithm: "ML-DSA-65".into(),
        timestamp,
        session_uuid: uuid.into(),
    };
    let pub_json = serde_json::to_string(&pub_obj).ok()?;
    let priv_json = serde_json::to_string(&priv_obj).ok()?;
    let mk: crate::r#type::MasterKey = serde_json::from_str(master_public_json).ok()?;
    let mh = key_hash(&mk.key);
    let sign = crate::master_key::sign_master_key(
        master_private_json,
        &pub_json,
        &mh,
    )?;
    Some((pub_json, priv_json, sign))
}

/// 秘密鍵 JSON の妥当性チェック
pub fn is_valid_identity_key_private(key_json: &str) -> bool {
    if let Ok(ik) = serde_json::from_str::<IdentityKey>(key_json) {
        ik.key_type == "identityKeyPrivate"
            && ik.algorithm == "ML-DSA-65"
            && is_valid_uuid_v7(&ik.session_uuid)
    } else { false }
}

/// 公開鍵 JSON の妥当性チェック
pub fn is_valid_identity_key_public(key_json: &str) -> bool {
    if let Ok(ik) = serde_json::from_str::<IdentityKey>(key_json) {
        ik.key_type == "identityKeyPublic"
            && ik.algorithm == "ML-DSA-65"
            && is_valid_uuid_v7(&ik.session_uuid)
    } else { false }
}

/// Sign JSON の妥当性チェック
pub fn is_valid_sign_identity_key(sign_json: &str) -> bool {
    if let Ok(sign) = serde_json::from_str::<crate::r#type::Sign>(sign_json) {
        sign.key_type == "identityKey"
            && sign.algorithm.as_deref() == Some("ML-DSA-65")
    } else { false }
}
